# SpaceX Launch Site Map

This notebook visualizes SpaceX launch sites and analyzes their proximity to relevant features like coastlines, railways, highways, and cities using Folium.

## Imports

This cell imports the necessary libraries for creating the interactive map and performing geospatial calculations:

In [11]:
# Import folium for creating interactive maps
import folium

# Import pandas for data manipulation
import pandas as pd

# Import the MarkerCluster plugin to group nearby points on the map
from folium.plugins import MarkerCluster

# Import the MousePosition plugin to display the coordinates of the cursor on the map
from folium.plugins import MousePosition

# Import DivIcon to add custom text labels or HTML elements to the map
from folium.features import DivIcon

*   **`folium`**: The main library used for creating interactive maps.
*   **`pandas`**: Used for data manipulation and analysis, particularly for reading and working with the launch site data.
*   **`folium.plugins.MarkerCluster`**: A plugin to group nearby markers on the map, improving performance and readability when dealing with many markers.
*   **`folium.plugins.MousePosition`**: A plugin to display the latitude and longitude of the mouse cursor on the map, useful for identifying coordinates.
*   **`folium.features.DivIcon`**: Used to create custom icons for markers using HTML, allowing for text labels on the map.

The subsequent cells will use these libraries to:

1.  Load and display the SpaceX launch site data.
2.  Visualize launch sites on a map.
3.  Calculate and visualize distances from launch sites to key geographic features.
4.  Analyze the strategic placement of launch sites based on these distances.

In [12]:
# define the world map centered around Canada with a low zoom level
world_map = folium.Map(location=[56.130, -106.35], zoom_start=4)

# display world map
world_map

## Initializing the Map

This cell initializes a Folium map object centered around Canada with a low zoom level. This map will be used to visualize the SpaceX launch sites and other relevant geographic features.

In [13]:
# Define the URL for the SpaceX launch geospatial dataset
URL = 'https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBM-DS0321EN-SkillsNetwork/datasets/spacex_launch_geo.csv'

# Read the CSV file from the URL into a pandas DataFrame
spacex_df = pd.read_csv(URL)

# Display the first few rows to inspect the data
display(spacex_df.head())

Flight Number        Date Time (UTC) Booster Version  Launch Site  \
0              1  2010-06-04   18:45:00  F9 v1.0  B0003  CCAFS LC-40   
1              2  2010-12-08   15:43:00  F9 v1.0  B0004  CCAFS LC-40   
2              3  2012-05-22    7:44:00  F9 v1.0  B0005  CCAFS LC-40   
3              4  2012-10-08    0:35:00  F9 v1.0  B0006  CCAFS LC-40   
4              5  2013-03-01   15:10:00  F9 v1.0  B0007  CCAFS LC-40   

                                             Payload  Payload Mass (kg)  \
0               Dragon Spacecraft Qualification Unit                0.0   
1  Dragon demo flight C1, two CubeSats,  barrel o...                0.0   
2                             Dragon demo flight C2+              525.0   
3                                       SpaceX CRS-1              500.0   
4                                       SpaceX CRS-2              677.0   

       Orbit         Customer        Landing Outcome  class        Lat  \
0        LEO           SpaceX  Failure   (parachute)      0  28.562302   
1  LEO (ISS)  NASA (COTS) NRO  Failure   (parachute)      0  28.562302   
2  LEO (ISS)      NASA (COTS)             No attempt      0  28.562302   
3  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   
4  LEO (ISS)       NASA (CRS)             No attempt      0  28.562302   

        Long  
0 -80.577356  
1 -80.577356  
2 -80.577356  
3 -80.577356  
4 -80.577356

## Loading and Displaying Data

This cell downloads the SpaceX launch data from a URL and loads it into a pandas DataFrame. It then displays the first few rows of the DataFrame to provide a preview of the data structure and content.

In [14]:
# Extract the launch site names along with their latitude and longitude coordinates
launch_sites_coords = spacex_df[['Launch Site', 'Lat', 'Long']]

# Print the coordinates of each launch site
print(launch_sites_coords)

     Launch Site        Lat        Long
0    CCAFS LC-40  28.562302  -80.577356
1    CCAFS LC-40  28.562302  -80.577356
2    CCAFS LC-40  28.562302  -80.577356
3    CCAFS LC-40  28.562302  -80.577356
4    CCAFS LC-40  28.562302  -80.577356
5    CCAFS LC-40  28.562302  -80.577356
6    CCAFS LC-40  28.562302  -80.577356
7    CCAFS LC-40  28.562302  -80.577356
8    CCAFS LC-40  28.562302  -80.577356
9    CCAFS LC-40  28.562302  -80.577356
10   CCAFS LC-40  28.562302  -80.577356
11   CCAFS LC-40  28.562302  -80.577356
12   CCAFS LC-40  28.562302  -80.577356
13   CCAFS LC-40  28.562302  -80.577356
14   CCAFS LC-40  28.562302  -80.577356
15   CCAFS LC-40  28.562302  -80.577356
16   CCAFS LC-40  28.562302  -80.577356
17   CCAFS LC-40  28.562302  -80.577356
18   CCAFS LC-40  28.562302  -80.577356
19   CCAFS LC-40  28.562302  -80.577356
20   CCAFS LC-40  28.562302  -80.577356
21   CCAFS LC-40  28.562302  -80.577356
22   CCAFS LC-40  28.562302  -80.577356
23   CCAFS LC-40  28.562302  -80.577356


## Extracting and Displaying Launch Site Coordinates

This cell extracts the `Launch Site`, `Lat`, and `Long` columns from the `spacex_df` DataFrame and stores them in a new DataFrame called `launch_sites_coords`. It then prints this new DataFrame, showing the latitude and longitude for each unique launch site.

In [15]:
# Set the initial center location of the map to NASA Johnson Space Center coordinates
nasa_coordinate = [29.559684888503615, -95.0830971930759]

# Create a Folium map centered at the NASA location with a starting zoom level of 10
site_map = folium.Map(location=nasa_coordinate, zoom_start=10)

## Initializing Map at NASA Johnson Space Center

This cell initializes a Folium map object centered around NASA Johnson Space Center in Houston, Texas. This map will be used as a starting point for visualizing the launch sites.

In [16]:
# Create a circle marker at NASA Johnson Space Center's coordinates
# The circle has a radius of 1000 meters, orange color, and includes a popup label
circle = folium.Circle(
    nasa_coordinate,
    radius=1000,
    color='#d35400',
    fill=True
).add_child(folium.Popup('NASA Johnson Space Center'))

# Create a text marker using DivIcon at the same location to label it on the map
marker = folium.map.Marker(
    nasa_coordinate,
    icon=DivIcon(
        icon_size=(20, 20),
        icon_anchor=(0, 0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % 'NASA JSC',
    )
)

# Add both the circle and the marker to the map
site_map.add_child(circle)
site_map.add_child(marker)

## Adding NASA Johnson Space Center to the Map

This cell adds a circle and a marker to the map at the coordinates of NASA Johnson Space Center. The circle has a radius of 1000 meters and is filled with blue. The marker is a text label displaying "NASA JSC". This helps to visualize the starting point of our analysis relative to the launch sites.

In [17]:
# Houston coordinate for map center and zoom reference
houston_coordinate = [29.7604, -95.3698]  # Latitude and Longitude of Houston

# Initialize the Folium map centered around Houston
site_map = folium.Map(location=houston_coordinate, zoom_start=5)

# Add a reference yellow circle for Houston
folium.Circle(
    location=houston_coordinate,
    radius=50000,  # radius in meters
    color='yellow',
    fill=True,
    fill_color='yellow',
    fill_opacity=0.5,
    popup='Houston'
).add_to(site_map)

# Loop through each launch site to add a circle and marker with its name
for idx, row in launch_sites_coords.iterrows():
    coordinate = [row['Lat'], row['Long']]

    # Add a blue circle for the launch site
    folium.Circle(
        location=coordinate,
        radius=1000,
        color='#000000',
        fill=True,
        fill_color='blue',
        fill_opacity=0.7,
        popup=row['Launch Site']
    ).add_to(site_map)

    # Add a text marker with the launch site name
    folium.map.Marker(
        location=coordinate,
        icon=DivIcon(
            icon_size=(150, 36),
            icon_anchor=(0, 0),
            html='<div style="font-size: 12; color:#d35400;"><b>%s</b></div>' % row['Launch Site']
        )
    ).add_to(site_map)

# Display the interactive map
site_map

## Visualizing Launch Sites

This cell initializes a Folium map centered around Houston, Texas, and then iterates through the launch site coordinates obtained earlier. For each launch site, it adds a blue circle with a radius of 1000 meters and a marker displaying the launch site name. This visualizes the locations of all the SpaceX launch sites on the map.

In [18]:
print(spacex_df.columns)

Index(['Flight Number', 'Date', 'Time (UTC)', 'Booster Version', 'Launch Site',
       'Payload', 'Payload Mass (kg)', 'Orbit', 'Customer', 'Landing Outcome',
       'class', 'Lat', 'Long'],
      dtype='object')


In [19]:
# Import folium if not already imported
import folium
from folium.features import DivIcon

# Color code: green for success, red for failure
def get_marker_color(outcome):
    if outcome == 1:
        return 'green'
    else:
        return 'red'

# Initial the map
site_map = folium.Map(location=[28.5, -80.6], zoom_start=5)  # approximate center for US launch sites

# Add a circle for each launch site (optional)
for idx, row in launch_sites_coords.iterrows():
    coordinate = [row['Lat'], row['Long']]
    folium.Circle(
        location=coordinate,
        radius=1000,
        color='#000000',
        fill=True,
        fill_color='blue',
        fill_opacity=0.7,
        popup=row['Launch Site']
    ).add_to(site_map)

# Add a marker for each launch showing success/failure
for idx, row in spacex_df.iterrows():
    coordinate = [row['Lat'], row['Long']]
    outcome_color = get_marker_color(row['class'])  # Class = 1 for success, 0 for failure

    folium.CircleMarker(
        location=coordinate,
        radius=5,
        color=outcome_color,
        fill=True,
        fill_color=outcome_color,
        fill_opacity=0.7,
        popup=f"Site: {row['Launch Site']}<br>Outcome: {'Success' if row['class']==1 else 'Failure'}"
    ).add_to(site_map)

# Display the map
site_map

## Visualizing Launch Outcomes

This cell adds markers to the map for each launch, color-coded based on the launch outcome. Green markers represent successful launches, and red markers represent failed launches. This allows for a visual analysis of launch success rates at different sites.

In [20]:
spacex_df.tail(10)

Flight Number        Date Time (UTC) Booster Version   Launch Site  \
46             43  2017-10-11   22:53:00  F9 FT  B1031.2    KSC LC-39A   
47             44  2017-10-30   19:34:00   F9 B4 B1042.1    KSC LC-39A   
48             54  2018-05-11   20:14:00  F9 B5  B1046.1    KSC LC-39A   
49             45  2017-12-15   15:36:00  F9 FT  B1035.2  CCAFS SLC-40   
50             47  2018-01-08    1:00:00   F9 B4 B1043.1  CCAFS SLC-40   
51             48  2018-01-31   21:25:00  F9 FT  B1032.2  CCAFS SLC-40   
52             50  2018-03-06    5:33:00     F9 B4 B1044  CCAFS SLC-40   
53             52  2018-04-02   20:30:00  F9 B4  B1039.2  CCAFS SLC-40   
54             53  2018-04-18   22:51:00   F9 B4 B1045.1  CCAFS SLC-40   
55             56  2018-06-04    4:45:00  F9 B4  B1040.2  CCAFS SLC-40   

                                         Payload  Payload Mass (kg)  \
46                         SES-11 / EchoStar 105            5200.00   
47                                   Koreasat 5A            3500.00   
48                                 Bangabandhu-1            3600.00   
49                                 SpaceX CRS-13            2205.00   
50                                          Zuma            3696.65   
51                             GovSat-1 / SES-16            4230.00   
52                        Hispasat 30W-6  PODSat            6092.00   
53                                 SpaceX CRS-14            2647.00   
54  Transiting Exoplanet Survey Satellite (TESS)             362.00   
55                                        SES-12            5384.00   

        Orbit             Customer       Landing Outcome  class        Lat  \
46        GTO         SES EchoStar  Success (drone ship)      1  28.573255   
47        GTO       KT Corporation  Success (drone ship)      1  28.573255   
48        GTO   Thales-Alenia/BTRC  Success (drone ship)      1  28.573255   
49  LEO (ISS)           NASA (CRS)  Success (ground pad)      1  28.563197   
50        LEO     Northrop Grumman  Success (ground pad)      1  28.563197   
51        GTO                  SES    Controlled (ocean)      0  28.563197   
52        GTO  Hispasat  NovaWurks            No attempt      0  28.563197   
53  LEO (ISS)           NASA (CRS)            No attempt      0  28.563197   
54        HEO           NASA (LSP)  Success (drone ship)      1  28.563197   
55        GTO                  SES            No attempt      0  28.563197   

         Long  
46 -80.646895  
47 -80.646895  
48 -80.646895  
49 -80.576820  
50 -80.576820  
51 -80.576820  
52 -80.576820  
53 -80.576820  
54 -80.576820  
55 -80.576820

In [21]:
marker_cluster = MarkerCluster()
# Create a new column 'marker_color'
spacex_df['marker_color'] = spacex_df['class'].apply(lambda x: 'green' if x == 1 else 'red')

# Check result
spacex_df[['class', 'marker_color']].head()

class marker_color
0      0          red
1      0          red
2      0          red
3      0          red
4      0          red

## Preparing Data for Clustered Markers

This cell initializes a `MarkerCluster` object to group nearby launch outcome markers, improving map performance and readability. It also creates a new column `marker_color` in the `spacex_df` DataFrame, assigning 'green' for successful launches (class 1) and 'red' for failed launches (class 0). This column will be used to color the markers in the cluster.

In [22]:
# Add marker_cluster to current site_map
site_map.add_child(marker_cluster)
for index, record in spacex_df.iterrows():
    marker_cluster.add_child(marker)
site_map

## Adding Clustered Markers to the Map

This cell adds the `marker_cluster` object, containing color-coded markers for each launch outcome, to the `site_map`. This groups nearby markers together, improving the visual clarity of the map, especially in areas with many launches.

In [23]:
# TASK 3: Calculate the distances between a launch site to its proximities
!pip install geopy

from geopy.distance import geodesic

# Example: let's say we have a launch site at (lat1, lon1)
launch_site = (28.562302, -80.577356)  # Example: CCAFS LC-40

# Example coordinates of nearby features
nearest_coastline = (28.563, -80.567)
nearest_railway = (28.572, -80.585)
nearest_highway = (28.563, -80.570)
nearest_city = (28.608, -80.807)   # Example: Titusville, FL

# Calculate distances
dist_coastline = geodesic(launch_site, nearest_coastline).km
dist_railway = geodesic(launch_site, nearest_railway).km
dist_highway = geodesic(launch_site, nearest_highway).km
dist_city = geodesic(launch_site, nearest_city).km

print(f"Distance to coastline: {dist_coastline:.2f} km")
print(f"Distance to railway: {dist_railway:.2f} km")
print(f"Distance to highway: {dist_highway:.2f} km")
print(f"Distance to city: {dist_city:.2f} km")

Distance to coastline: 1.02 km
Distance to railway: 1.31 km
Distance to highway: 0.72 km
Distance to city: 23.03 km


## Calculating Distances to Proximities

This cell demonstrates how to calculate the distance between a launch site and various nearby features like the coastline, railway, highway, and a city using the `geodesic` function from the `geopy` library. This is a crucial step in analyzing the strategic placement of launch sites.

In [24]:
# Add Mouse Position to get the coordinate (Lat, Long) for a mouse over on the map
formatter = "function(num) {return L.Util.formatNum(num, 5);};"
mouse_position = MousePosition(
    position='topright',
    separator=' Long: ',
    empty_string='NaN',
    lng_first=False,
    num_digits=20,
    prefix='Lat:',
    lat_formatter=formatter,
    lng_formatter=formatter,
)

site_map.add_child(mouse_position)
site_map

## Adding Mouse Position to the Map

This cell adds the `MousePosition` plugin to the map. This plugin displays the latitude and longitude of the mouse cursor as you hover over the map, which is useful for obtaining coordinates of specific points of interest.

In [25]:
from math import sin, cos, sqrt, atan2, radians

def calculate_distance(lat1, lon1, lat2, lon2):
    # approximate radius of earth in km
    R = 6373.0

    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c
    return distance

## Distance Calculation Function

This cell defines a function `calculate_distance` that calculates the distance between two geographical points (specified by their latitude and longitude) using the Haversine formula. This function will be used in subsequent steps to determine the distances between launch sites and nearby features.

In [26]:
import numpy as np

# Define Haversine distance function
def calculate_distance(lat1, lon1, lat2, lon2):
    # Convert degrees to radians
    lat1, lon1, lat2, lon2 = map(np.radians, [lat1, lon1, lat2, lon2])

    # Haversine formula
    dlat = lat2 - lat1
    dlon = lon2 - lon1
    a = np.sin(dlat/2)**2 + np.cos(lat1)*np.cos(lat2)*np.sin(dlon/2)**2
    c = 2 * np.arcsin(np.sqrt(a))
    R = 6371  # Earth radius in km
    return R * c

# Launch site coordinates (CCAFS LC-40)
launch_site_lat, launch_site_lon = 28.562302, -80.577356

# Coastline coordinates (picked using MousePosition)
coastline_lat, coastline_lon = 28.56367, -80.57163

# Calculate distance
distance_coastline = calculate_distance(launch_site_lat, launch_site_lon,
                                        coastline_lat, coastline_lon)

print("Distance from launch site to closest coastline:", round(distance_coastline, 2), "km")


Distance from launch site to closest coastline: 0.58 km


## Calculating Distance to Closest Coastline

This cell calculates the distance between a specific launch site (CCAFS LC-40 in this example) and a manually identified point on the closest coastline using the `calculate_distance` function defined earlier. The calculated distance is then printed.

**Result:** The distance from the launch site to the closest coastline is approximately 0.58 km.

In [27]:
launch_site = [28.562302, -80.577356]   # CCAFS LC-40
coastline = [28.56367, -80.57163]       # Closest coastline point

# Calculate distance using your function
distance = calculate_distance(launch_site[0], launch_site[1],
                              coastline[0], coastline[1])

# Add marker for the coastline point
folium.Marker(
    location=coastline,
    popup="Closest Coastline",
    icon=folium.Icon(color='blue', icon='info-sign')
).add_to(world_map)

# Add distance label as text on the map
distance_marker = folium.Marker(
    location=coastline,
    icon=DivIcon(
        icon_size=(20,20),
        icon_anchor=(0,0),
        html='<div style="font-size: 12px; color:#d35400;"><b>%s</b></div>' % ("{:.2f} KM".format(distance)),
    )
).add_to(world_map)

# Show map
world_map


## Visualizing Distance to Closest Coastline

This cell adds a marker at the manually identified closest coastline point and displays the calculated distance from the launch site to this point on the map. This helps visualize the proximity of the launch site to the coast.

In [28]:
import folium

# Initialize the map centered near the launch site
world_map = folium.Map(location=launch_site, zoom_start=10)

# Launch site and closest coastline point coordinates
launch_site = [28.562302, -80.577356]  # Example: CCAFS LC-40
coastline = [28.56367, -80.57163]      # Closest coastline point

# Draw a PolyLine between the launch site and the coastline
lines = folium.PolyLine(
    locations=[launch_site, coastline],
    weight=2,          # thickness of the line
    color='blue',      # line color
    opacity=0.6
)

# Add the line to the map
world_map.add_child(lines)

# Optional: Add markers for launch site and coastline
folium.Marker(launch_site, popup='Launch Site').add_to(world_map)
folium.Marker(coastline, popup='Coastline Point').add_to(world_map)

# Display the map
world_map


## Visualizing Distance to Closest Coastline

This cell initializes a Folium map centered near the launch site and draws a blue line connecting the launch site to the manually identified closest coastline point. This visually represents the distance between the two locations.

In [29]:
# Draw a line between the marker to the launch site
from geopy.distance import geodesic
import folium
from folium.features import DivIcon

# Launch site coordinates
launch_site = [28.562302, -80.577356]  # Example: CCAFS LC-40

# Example: closest city coordinates
city = [28.5721, -80.5856]  # Replace with actual coordinates

# Calculate distance
distance_city = geodesic(launch_site, city).km

# Add marker for city
city_marker = folium.Marker(
    city,
    icon=DivIcon(
        icon_size=(150,36),
        icon_anchor=(0,0),
        html='<div style="font-size: 12; color:#d35400;"><b>%s KM</b></div>' % "{:.2f}".format(distance_city)
    )
)
world_map.add_child(city_marker)
# Draw line from launch site to city
line_to_city = folium.PolyLine(
    locations=[launch_site, city],
    weight=2,
    color='red',
    opacity=0.6
)
world_map.add_child(line_to_city)

# Display map
world_map

## Visualizing Distance to Closest City

This cell calculates the distance between a launch site and a manually identified closest city. It then adds a marker for the city and draws a red line connecting the launch site to the city on the map, along with a label showing the calculated distance. This visualizes the proximity of the launch site to a major urban area.

Based on the distances we calculated and visualized on the map, here are the answers to your questions:

1-Are launch sites in close proximity to railways?

Yes, based on the example calculation for CCAFS LC-40, the distance to the nearest railway is relatively small (around 1.31 km). This suggests that launch sites are often located near railways, likely for transporting large components.

2-Are launch sites in close proximity to highways?

Yes, the example for CCAFS LC-40 shows a very small distance to the nearest highway (around 0.72 km). This is expected as highways are crucial for accessibility and transporting personnel and equipment.

3-Are launch sites in close proximity to coastline?

Yes, the example for CCAFS LC-40 shows a distance of less than a kilometer to the coastline (around 0.58 km). Launch sites are often located near coastlines for safety reasons, allowing rockets to launch over the ocean and minimizing risk to populated areas in case of a failure.

4-Do launch sites keep certain distance away from cities?

Yes, the example for CCAFS LC-40 shows a distance of over 23 km to the nearest city (Titusville, FL). This supports the idea that launch sites are situated away from densely populated areas to ensure public safety.
Explanation of Findings:

The locations of SpaceX launch sites appear to be strategically chosen to balance logistical needs with safety considerations. Their close proximity to railways and highways facilitates the transportation of rocket components and personnel. The proximity to the coastline is a critical safety measure, allowing launches to occur over the ocean. Finally, their distance from major cities is likely a deliberate choice to minimize risk to the public in the event of an anomaly during launch.